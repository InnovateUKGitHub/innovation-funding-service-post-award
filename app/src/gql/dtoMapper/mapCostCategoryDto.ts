import { PCROrganisationType } from "@framework/constants";
import type { CostCategoryDto } from "@framework/dtos";
import { costCategoryTypeMapper } from "@framework/mappers/costCategory";
import { numberComparator } from "@framework/util";

/**
 * from the root `Acc_CostCategory__c` node
 */
type CostCategoryNode = Readonly<
  Partial<{
    Id: string;
    Acc_CostCategoryName__c: GQL.Value<string>;
    Acc_OrganisationType__c: GQL.Value<string>;
    Acc_CompetitionType__c: GQL.Value<string>;
    Acc_DisplayOrder__c: GQL.Value<number>;
  }>
> | null;

/**
 * from the root `Acc_Profile__c` node
 */
type ProfileNodeForRequiredCostCategories = ReadonlyArray<{
  node: {
    Acc_CostCategory__c: GQL.Value<string>;
  } | null;
} | null> | null;

type CostCategoryDtoMapping = Pick<
  CostCategoryDto,
  "id" | "competitionType" | "name" | "isCalculated" | "organisationType" | "type"
> & { displayOrder: number };

const mapper: GQL.DtoMapper<CostCategoryDtoMapping, CostCategoryNode> = {
  id: function (node) {
    return node?.Id ?? "";
  },
  name: function (node) {
    return node?.Acc_CostCategoryName__c?.value ?? "";
  },
  type: function (node) {
    return costCategoryTypeMapper(
      node?.Acc_OrganisationType__c?.value ?? "Unknown",
      node?.Acc_CostCategoryName__c?.value ?? "unknown",
    );
  },
  competitionType: function (node) {
    return node?.Acc_CompetitionType__c?.value ?? "unknown";
  },
  organisationType: function (node) {
    return (node?.Acc_OrganisationType__c?.value ?? "unknown") as PCROrganisationType;
  },
  isCalculated: function () {
    return false;
  },
  displayOrder: function (node) {
    return node?.Acc_DisplayOrder__c?.value ?? 0;
  },
};

/**
 * Maps a specified CostCategory Node from a GQL query to a slice of
 * the CostCategoryDto to ensure consistency and compatibility in the application
 */
export function mapToCostCategoryDto<T extends CostCategoryNode, PickList extends keyof CostCategoryDtoMapping>(
  node: T,
  pickList: PickList[],
): Pick<CostCategoryDtoMapping, PickList> {
  return pickList.reduce((dto, field) => {
    dto[field] = mapper[field](node);
    return dto;
  }, {} as Pick<CostCategoryDtoMapping, PickList>);
}

/**
 * Maps CostCategory Edge to array of Cost Category DTOs.
 */
export function mapToCostCategoryDtoArray<
  T extends ReadonlyArray<{ node: CostCategoryNode } | null> | null,
  PickList extends keyof CostCategoryDtoMapping,
>(edges: T, pickList: PickList[]): Pick<CostCategoryDtoMapping, PickList>[] {
  return (
    edges?.map(node => {
      return mapToCostCategoryDto(node?.node ?? null, pickList);
    }) ?? []
  );
}

/**
 * Maps CostCategory Edge to filtered and sorted list of cost categories.
 *
 * `id`, `name` and `displayOrder` are required fields as is an array of requiredCategory Ids
 *
 * This also requires the `Acc_Profile__c` root node containing `Acc_CostCategory__c`
 */
export function mapToRequiredSortedCostCategoryDtoArray<
  T extends ReadonlyArray<{ node: CostCategoryNode } | null> | null,
  PickList extends keyof CostCategoryDtoMapping,
>(
  edges: T,
  pickList: [...["id", "name", "displayOrder"], ...PickList[]],
  profileNode: ProfileNodeForRequiredCostCategories,
): Pick<CostCategoryDtoMapping, "id" | "name" | "displayOrder" | PickList>[] {
  return (edges ?? [])
    .map(node => {
      return mapToCostCategoryDto(node?.node ?? null, pickList);
    })
    .filter(
      x =>
        !!x?.name &&
        (profileNode ?? []).some(
          profile => (profile?.node?.Acc_CostCategory__c?.value ?? "unknown category id") === x.id,
        ),
    )
    .sort((a, b) => numberComparator(a.displayOrder, b.displayOrder));
}
