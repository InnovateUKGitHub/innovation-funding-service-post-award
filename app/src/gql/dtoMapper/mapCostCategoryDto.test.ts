import { mapToRequiredSortedCostCategoryDtoArray } from "./mapCostCategoryDto";

describe("mapToRequiredSortedCostCategoryDtoArray", () => {
  const edges = [
    {
      node: {
        Id: "0",
        Acc_CostCategoryName__c: { value: "Labour" },
        Acc_OrganisationType__c: { value: "Academic" },
        Acc_CompetitionType__c: { value: "CR&D" },
        Acc_DisplayOrder__c: { value: 3 },
      },
    },
    {
      node: {
        Id: "1",
        Acc_CostCategoryName__c: { value: "Travel" },
        Acc_OrganisationType__c: { value: "Academic" },
        Acc_CompetitionType__c: { value: "CR&D" },
        Acc_DisplayOrder__c: { value: 2 },
      },
    },
    {
      node: {
        Id: "2",
        Acc_CostCategoryName__c: { value: "Materials" },
        Acc_OrganisationType__c: { value: "Business" },
        Acc_CompetitionType__c: { value: "KTP" },
        Acc_DisplayOrder__c: { value: 1 },
      },
    },
    {
      node: {
        Id: "3",
        Acc_CostCategoryName__c: { value: "Other costs" },
        Acc_OrganisationType__c: { value: "Business" },
        Acc_CompetitionType__c: { value: "Loans" },
        Acc_DisplayOrder__c: { value: 4 },
      },
    },
    {
      node: {
        Id: "4",
        Acc_CostCategoryName__c: { value: "Bananas" },
        Acc_OrganisationType__c: { value: "Business" },
        Acc_CompetitionType__c: { value: "CR&D" },
        Acc_DisplayOrder__c: { value: 5 },
      },
    },
    {
      node: {
        Id: "5",
        Acc_CostCategoryName__c: { value: "Schmoozing" },
        Acc_OrganisationType__c: { value: "Business" },
        Acc_CompetitionType__c: { value: "CR&D" },
        Acc_DisplayOrder__c: { value: 6 },
      },
    },
    {
      node: {
        Id: "6",
        Acc_CostCategoryName__c: { value: "Kickbacks" },
        Acc_OrganisationType__c: { value: "Business" },
        Acc_CompetitionType__c: { value: "CR&D" },
        Acc_DisplayOrder__c: { value: 7 },
      },
    },
  ];

  const profileNode = ["0", "1s", "2", "3"].map(x => ({
    node: {
      Acc_CostCategory__c: {
        value: x,
      },
    },
  }));

  it("should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records", () => {
    expect(
      mapToRequiredSortedCostCategoryDtoArray(
        edges,
        ["id", "name", "displayOrder", "type", "competitionType", "organisationType", "isCalculated"],
        profileNode,
      ),
    ).toMatchSnapshot();
  });

  it("should map the gql data only including fields on the picklist", () => {
    expect(
      mapToRequiredSortedCostCategoryDtoArray(edges, ["id", "name", "displayOrder", "type"], profileNode),
    ).toMatchSnapshot();
  });
});
