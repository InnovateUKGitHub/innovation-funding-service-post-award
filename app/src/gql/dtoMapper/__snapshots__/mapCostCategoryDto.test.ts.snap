// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToRequiredSortedCostCategoryDtoArray should map the gql data only including fields on the picklist 1`] = `
Array [
  Object {
    "displayOrder": 1,
    "id": "2",
    "name": "Materials",
    "type": 6,
  },
  Object {
    "displayOrder": 3,
    "id": "0",
    "name": "Labour",
    "type": 2,
  },
  Object {
    "displayOrder": 4,
    "id": "3",
    "name": "Other costs",
    "type": 10,
  },
]
`;

exports[`mapToRequiredSortedCostCategoryDtoArray should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records 1`] = `
Array [
  Object {
    "competitionType": "KTP",
    "displayOrder": 1,
    "id": "2",
    "isCalculated": false,
    "name": "Materials",
    "organisationType": "Business",
    "type": 6,
  },
  Object {
    "competitionType": "CR&D",
    "displayOrder": 3,
    "id": "0",
    "isCalculated": false,
    "name": "Labour",
    "organisationType": "Academic",
    "type": 2,
  },
  Object {
    "competitionType": "Loans",
    "displayOrder": 4,
    "id": "3",
    "isCalculated": false,
    "name": "Other costs",
    "organisationType": "Business",
    "type": 10,
  },
]
`;
